### ISSUE TITLE
Proposal addition to object-shorthand: add consistent and consistent-as-needed

### ISSUE THREAD

**Issue by martijndeh** [eslint/eslint#5438]
**When does this rule warn? Please describe and show example code:**
Similar to the `quote-props` rule, where it's possible to set the rule to `consistent` and `consistent-as-needed`, I'd like the `object-shorthand` rule to include these options as well.

When `object-shorthand` set to `consistent`, the below should not be considered a problem:
```js
let obj1 = {
    foo: foo,
    bar: 'This is a string',
};
```
No problem:
```js
let obj2 = {
    foo,
    bar,
};
```

The following should be considered a problem:
```js
let obj1 = {
    foo,
    bar: 'This is a string',
};
```

Problem:
```js
let obj2 = {
    a() {},
    bar: foo,
};
```

The `consistent-as-needed` is very similar, but if all keys are redundant, the following pattern is considered a problem:
```js
let obj3 = {
    foo: foo,
    bar: bar,
};
```

The options apply to both properties and methods.

**Is this rule preventing an error or is it stylistic?**
Stylistic. Personally, I feel the shorthand syntax is great, but mixing it with the longform syntax makes it harder to scan/read code. In my rules, I want to allow the shorthand syntax, but only if it's consistent in that object literal.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**
I think this is in line with other rules which also provide consistent and consistent-as-needed options (e.g. `quote-props`).

**Are you willing to create the rule yourself?**
Yes.

**Comment by platinumazure** [comment]
Maybe it's just that I don't do ES6 much, but I'm having a hard time distinguishing what is "shorthand" and what is not.

Can you tell me which of the properties below (1) do *you* consider "shorthand", and (2) does the *rule* consider "shorthand"? Thanks!

```js
var obj = {
    foo,
    bar: bar,
    a: function () {},
    function a() {},
    get a() {},
    set a(value) {}
};
```

**Comment by martijndeh** [comment]
Sure. Please see the below annotated object literal:

```js
var obj = {
    // This is shorthand and new syntax in ES6. The key is implicit. This was invalid in ES5.
    foo, 

    // This is longform as you need to explicitly state the key. This was valid in ES5.
    bar: bar, 

    // This is longform. You need to explicitly state the key. This was valid in ES5.
    a: function () {}, 

    // This is shorthand. This was invalid in ES5. The key is implicit.
    function a() {}, 

    // This is a more ambiguous case. It's not included in the new object literal 
    // shorthand syntax as introduced in ES6, but it also doesn't explicitly state the key (it's 
    // very similar to the named function above).
    //
    // Personally, I consider this neither longform nor shorthand.
    get a() {},

    // Idem to the getter.
    set a(value) {} 
};
```

I have a PR coming up (see https://github.com/martijndeh/eslint/tree/feature/object-shorthand-consistent) which implements this feature. Personally I hardly use getters and setts inside object literals, so I decided to exclude these expressions from the consistency check.

This means the following is valid when setting the `"consistent"` option:

```js
var obj = {
    foo,
    bar,
    get test() { return 1; }
};
```

But, the following is valid as well:

```js
var obj = {
    foo: foo,
    bar: bar,
    get test() { return 1; }
};
```

When setting the `"consistent-as-needed"` option the following would be invalid:

```js
var obj = {
    foo: foo, // should be foo
    bar: bar, // should be bar
    get test() { return 1; } // ignored
};
```

What do you think?

**Comment by platinumazure** [comment]
Thanks. FYI, we usually don't merge pull requests until the team has fully evaluated the issue and marked it as "accepted", so don't put too much time into polishing up the pull request (otherwise that time would be wasted if we reject the request). See [this page](http://eslint.org/docs/developer-guide/contributing/working-on-issues#issue-labels) for more info.

**Comment by michaelficarra** [comment]
You both seem to be confused about how to write methods.

```js
({ a(){} });
```

**Comment by martijndeh** [comment]
Ah my bad. I shouldn't have included the `function` keyword in the shorthand function example in the original example. Edited.

**Comment by platinumazure** [comment]
@michaelficarra I did note I don't use or know ES6, hence asking the
question in the first place. Thank you for the education.

**Comment by BYK** [comment]
I think we just need `always-consistent` or something to enforce consistency for cases when we cannot use the shorthand notation.

**Comment by martijndeh** [comment]
That's `consistent-as-needed`, right?

**Comment by BYK** [comment]
It's not really "as-needed". I mean "as-needed" as in "we need to **not** use the shorthand syntax", hence the confusion.

**Comment by martijndeh** [comment]
Can you give some examples when `always-consistent` would apply and not apply? You are suggesting it as an option in addition to the other options in the proposal, correct?

**Comment by nzakas** [comment]
@BYK I think the point is to keep the naming similar to existing options.

@eslint/eslint-team this proposal has been largely ignored. We need a champion and three :+1:s to continue or else we should close this.

**Comment by BYK** [comment]
I guess that's fair.

**Comment by nzakas** [comment]
Sorry, lost track of this. @martijndeh are you still interested?

I'll champion this change since there's a PR already.

**Comment by martijndeh** [comment]
No worries. Yes. Personally I really want this rule. If there is support I'll make sure the PR gets fixed.

**Comment by nzakas** [comment]
@eslint/eslint-team can I get some 👍s for this? I think it makes sense and the PR is basically already done.

**Comment by gyandeeps** [comment]
:+1:

**Comment by mikesherov** [comment]
👍

**Comment by mysticatea** [comment]
Sounds good to me :+1:

**Comment by nzakas** [comment]
Okay @martijndeh we are good to move forward. Sorry for the delay.
